src/useFish.ts

```diff
@@ -13,176 +13,124 @@
-  stream$: Observable<PublicState>
+  state: State
+  /** run function for the observed fish */
+  run: (eff: StateEffect<State, Event>) => Promise<void>
+  /** current props of the fish */
+  props?: T
 }
+export type SetProps<P> = (props: P) => void
 
-/**
- * Stateful integration of an actyx Pond Fish.
- *
- * ## Example:
- * ```js
- * export const App = () => {
- *   const [chatRoomFish] = useFish(ChatRoomFish, 'lobby')
- *
- *   return (
- *     <div>
- *       {chatRoomFish && (
- *         <>
- *           <div>current chat room: {chatRoomFish.name}</div>
- *           <div>
- *             {chatRoomFish.state.map((message, idx) => <div key={idx}>{message}</div>)}
- *           </div>
- *           <div>
- *             <button
- *               onClick={
- *                 () => chatRoomFish.feed({type: CommandType.postMessage, sender: 'me', 'hi'})
- *               }
- *             >
- *               send
- *             </button>
- *           </div>
- *         </>
- *       )}
- *     </div>
- *   )
- * }
- * ```
- *
- * @param fish fish to get the public state for
- * @param fishName (optional) if a name is set, this fish will be observed or hydrated. If no name is set, use the returned SetNewFishName to set the name
- * @typeParam Command Available commands for the fish
- * @typeParam PublicState Observable state of the fish
- * @returns [ReactFish?, SetNewFishName] If no name is set, the ReactFish is undefined
- */
-export const useFish = <Command, PublicState>(
-  fish: FishTypeEntity<Command, PublicState>,
-  fishName?: string
-): [ReactFish<Command, PublicState> | undefined, SetNewFishName] => {
-  const [name, setName] = React.useState(fishName)
-  const [reactFish, setReactFish] = React.useState<ReactFish<Command, PublicState>>()
-  const { observe, feed } = usePond()
+/** @internal */
+const runStateEffect = <S, E>(pond: Pond, fish: Fish<S, E>) => (eff: StateEffect<S, E>) =>
+  pond.run(fish, eff).toPromise()
 
-  React.useEffect(() => {
-    if (name) {
-      const stream$ = observe(fish, name).share()
+/** @internal */
+const mkReactFish = <S, E, P>(
+  pond: Pond,
+  fish: Fish<S, E>,
+  state: S,
+  props?: P
+): ReactFish<S, E, P> => ({
+  state,
+  run: runStateEffect(pond, fish),
+  props
+})
+
+/** @internal */
+const useFishFn = <State, E, P>(
+  mkFish: (props: P) => Fish<State, E>,
+  props?: P
+): [ReactFish<State, E, P> | undefined, SetProps<P>] => {
+  const pond = usePond()
+  const [fishProps, setFishProps] = React.useState<P | undefined>(props)
+  const [reactFish, setReactFish] = React.useState<ReactFish<State, E, P>>()
 
-      const sub = stream$
-        .pipe(
-          map<PublicState, ReactFish<Command, PublicState>>(state => ({
-            state,
-            feed: (command: Command) => feed(fish, name)(command).toPromise(),
-            name,
-            stream$
-          }))
-        )
-        .subscribe(setReactFish)
-      return () => sub && sub.unsubscribe()
+  React.useEffect(() => {
```

Comment:
this needs to cancel previous subscriptions, otherwise we leak memory

Reply:
we `return sub`. React will make sure to call this function before executing this closure again.
https://github.com/actyx-contrib/react-pond/pull/1/files#diff-43583109cd9f4bec37c4d5d5ee47fe93R67

Reply:
ah, okay, thanks for the explanation!